pkg/provider/azure_local_services.go

```diff
@@ -218,3 +275,271 @@ func newBatchOperationResult(name string, success bool, err error) batchOperatio
+	if ipv6 {
+		return fmt.Sprintf("%s-ipv6", serviceName)
+	}
+	return serviceName
+}
+
+// getBackendPoolNameForService determine the expected backend pool name
+// by checking the external traffic policy of the service.
+func (az *Cloud) getBackendPoolNameForService(service *v1.Service, clusterName string, ipv6 bool) string {
+	if !isLocalService(service) || !az.useMultipleStandardLoadBalancers() {
+		return getBackendPoolName(clusterName, ipv6)
+	}
+	return getLocalServiceBackendPoolName(getServiceName(service), ipv6)
+}
+
+// getBackendPoolNamesForService determine the expected backend pool names
+// by checking the external traffic policy of the service.
+func (az *Cloud) getBackendPoolNamesForService(service *v1.Service, clusterName string) map[bool]string {
+	if !isLocalService(service) || !az.useMultipleStandardLoadBalancers() {
+		return getBackendPoolNames(clusterName)
+	}
+	return map[bool]string{
+		consts.IPVersionIPv4: getLocalServiceBackendPoolName(getServiceName(service), false),
+		consts.IPVersionIPv6: getLocalServiceBackendPoolName(getServiceName(service), true),
+	}
+}
+
+// getBackendPoolIDsForService determine the expected backend pool IDs
+// by checking the external traffic policy of the service.
+func (az *Cloud) getBackendPoolIDsForService(service *v1.Service, clusterName, lbName string) map[bool]string {
+	if !isLocalService(service) || !az.useMultipleStandardLoadBalancers() {
+		return az.getBackendPoolIDs(clusterName, lbName)
+	}
+	return map[bool]string{
+		consts.IPVersionIPv4: az.getLocalServiceBackendPoolID(getServiceName(service), lbName, false),
+		consts.IPVersionIPv6: az.getLocalServiceBackendPoolID(getServiceName(service), lbName, true),
+	}
+}
+
+// getLocalServiceBackendPoolID gets the ID of the backend pool of a local service.
+func (az *Cloud) getLocalServiceBackendPoolID(serviceName string, lbName string, ipv6 bool) string {
+	return az.getBackendPoolID(lbName, getLocalServiceBackendPoolName(serviceName, ipv6))
+}
+
+// localServiceOwnsBackendPool checks if a backend pool is owned by a local service.
+func localServiceOwnsBackendPool(serviceName, bpName string) bool {
+	prefix := strings.Replace(serviceName, "/", "-", -1)
+	return strings.HasPrefix(strings.ToLower(bpName), strings.ToLower(prefix))
+}
+
+type serviceInfo struct {
+	ipFamily string
+	lbName   string
+}
+
+func newServiceInfo(ipFamily, lbName string) *serviceInfo {
+	return &serviceInfo{
+		ipFamily: ipFamily,
+		lbName:   lbName,
+	}
+}
+
+// getLocalServiceEndpointsNodeNames gets the node names that host all endpoints of the local service.
+func (az *Cloud) getLocalServiceEndpointsNodeNames(service *v1.Service) (sets.Set[string], error) {
+	var ep *discovery_v1.EndpointSlice
+	az.endpointSlicesCache.Range(func(key, value interface{}) bool {
+		endpointSlice := value.(*discovery_v1.EndpointSlice)
+		if strings.EqualFold(getServiceNameOfEndpointSlice(endpointSlice), service.Name) {
```

Comment:
Need to check service namespace also

Reply:
Thanks for catching, please check the new commit.